modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/coercion/BaseTypeCheckExecutionTest.java

```diff
@@ -0,0 +1,223 @@
+
+                BigDecimal bd = new BigDecimal(numericBase);
+                return bd.setScale(scale, RoundingMode.UNNECESSARY);
+            default:
+                throw new AssertionError("Unexpected type: " + type0);
+        }
+    }
+
+    private static Object generateConstantValueByType(NativeType type) {
+        return generateConstantValueByType(type, "2");
+    }
+
+    static ClusterWrapper testCluster(TypePair typePair, DataProvider<Object[]> dataProvider) {
+        TestCluster cluster = TestBuilders.cluster().nodes("N1")
+                .addTable().name("T")
+                .addKeyColumn("id", NativeTypes.INT32)
+                .addColumn("C1", typePair.first())
+                .addColumn("C2", typePair.second())
+                .end()
+                .dataProvider("N1", "T", TestBuilders.tableScan(dataProvider))
+                .build();
+
+        return new ClusterWrapper(cluster);
+    }
+
+    static class ClusterWrapper implements AutoCloseable {
+        private final TestCluster cluster;
+
+        ClusterWrapper(TestCluster cluster) {
+            this.cluster = cluster;
+
+            cluster.start();
+        }
+
+        void process(String sql, Matcher<Object> resultMatcher) {
+            var gatewayNode = cluster.node("N1");
+            var plan = gatewayNode.prepare(sql);
+            ResultSetMetadata resultMeta = plan.metadata();
+            ColumnMetadata colMeta = resultMeta.columns().get(0);
+
+            for (InternalSqlRow row : CursorUtils.getAllFromCursor(gatewayNode.executePlan(plan))) {
+                assertNotNull(row);
+                assertNotNull(row.get(0), "Await not null object");
+                assertThat(new Pair<>(row.get(0), colMeta), resultMatcher);
+            }
+        }
+
+        @Override
+        public void close() throws Exception {
+            cluster.stop();
+        }
+    }
+
+    /** Return results matcher, compare return type, precision and scale with analyzed object. */
+    static Matcher<Object> checkReturnResult() {
+        return new BaseMatcher<>() {
+            Object result;
+            ColumnMetadata meta;
+
+            @Override
+            public boolean matches(Object actual) {
+                assert actual != null;
+
+                Pair<Object, ColumnMetadata> pair = (Pair<Object, ColumnMetadata>) actual;
+
+                result = pair.getFirst();
+                meta = pair.getSecond();
+
+                int precision = 0;
+                int scale = 0;
+
+                if (result instanceof BigDecimal) {
+                    precision = ((BigDecimal) result).precision();
+                    scale = ((BigDecimal) result).scale();
+                }
+
+                boolean checkPrecisionScale = (result.getClass() != Float.class)
+                        && (result.getClass() != Double.class)
+                        && (result.getClass() != Boolean.class);
```

Comment:
```suggestion
                boolean checkPrecisionScale = result.getClass().isAssignableFrom(BigDecimal.class);
```

Reply:
or better just say that `checkPrecisionScale=true' inside of previous `if` statement

Reply:
disagree, in such a case you will miss comparison if return result is integer and, for example, metadata return wrong negative result

Reply:
Hm.... I'm just wondering why we have a negative scale for these types and why we have 0 for others.
Is it explainable or can it be considered as a bug?

Reply:
this is our placeholder for undefined scale ColumnMetadata#UNDEFINED_SCALE